ssl/ssl_sess.c

```diff
@@ -156,16 +156,12 @@ SSL_SESSION *SSL_get1_session(SSL *ssl)
 /* variant of SSL_get_session: caller really gets something */
 {
     SSL_SESSION *sess;
-    /*
-     * Need to lock this all up rather than just use CRYPTO_add so that
-     * somebody doesn't free ssl->session between when we check it's non-null
-     * and when we up the reference count.
-     */
-    CRYPTO_w_lock(CRYPTO_LOCK_SSL_SESSION);
+
     sess = ssl->session;
+    CRYPTO_MUTEX_lock_write(&sess->lock);
```

Comment:
> Why is the lock moved to after sess is assigned

Because sess is an invalid pointer until it gets initialized..

> it can be thread unsafe now

The problem really is that sess can be NULL, so that new code is wrong...